### ISSUE TITLE
OpenVPN process status says not running when it is.

### ISSUE THREAD

**Issue by marjohn56** [opnsense/core#3223]
https://forum.opnsense.org/index.php?topic=11562.0

This seems to be a stick pid issue. We are testing a solution for it.

**Comment by AdSchellevis** [comment]
the status check connects to the socket, if that doesn't work, this might happen. (I've seen something similar over here too)

https://github.com/opnsense/core/blob/d27cc8364252309c8e701bb805fbd8545f53ab50/src/etc/inc/plugins.inc.d/openvpn.inc#L1304

If it reports an error, we might want to check for a running pid before returning not active, if I'm not mistaken, you can't restart the server in these cases (since reported running), right?

**Comment by marjohn56** [comment]
The pid's don't match. The running process pid and the one in var/run are different. Thus the dashboard says nay but openvpn says yay! 

You can't restart no, as the pid is wrong. Manually killing it will not help either as you also need to delete the stuck pid,  do that and all is good. What I've done in my test is to force a clean up of the stuck pid in the openvpn restart function, at present it just deletes the stuck pid. What really should be done is to check that the process is actually not running - THEN delete the pid, however lets see what happens with my tester and see what he reports.

**Comment by AdSchellevis** [comment]
sounds like a plan :)

**Comment by fichtner** [comment]
From previous tickets this seems to be a race in OpenVPN or maybe in the way we restart it in an edge case. However, it was not reproducible from my end.

**Comment by marjohn56** [comment]
No and I cannot reproduce it either, or at least only on rare occasions. Ned however can reproduce it at will. He's had the issue since the early 18 series, but now it's starting to annoy him so I said I'd tale a look.

I think it only happens on restart, not sure if it does it on IF down/up. Whichever way this occurs, the only usual way a process would leave its pid behind is if it exits abnormally.

**Comment by Space2Man** [comment]
Hi,

some things I found out. This happens always after a restart:

```
root@OPNvirt:~ # ls -lT /var/run/openvpn_server1.pid 
-rw-r--r--  1 root  wheel  6 Feb 12 20:31:58 2019 /var/run/openvpn_server1.pid
```

After reboot I see three times a Resync

```
Feb 12 20:31:45 OPNvirt opnsense: /usr/local/etc/rc.openvpn: OpenVPN: One or more OpenVPN tunnel endpoints may have changed its IP. Reloading endpoints that may use WAN_DHCP6.
Feb 12 20:31:45 OPNvirt opnsense: /usr/local/etc/rc.openvpn: OpenVPN: Resync server1 SpaceNet OpenVPN Server
Feb 12 20:31:47 OPNvirt opnsense: /usr/local/etc/rc.openvpn: OpenVPN: One or more OpenVPN tunnel endpoints may have changed its IP. Reloading endpoints that may use WAN_IPv4_Dokom21.
Feb 12 20:31:51 OPNvirt opnsense: /usr/local/etc/rc.openvpn: OpenVPN: Resync server1 SpaceNet OpenVPN Server
Feb 12 20:31:57 OPNvirt opnsense: /usr/local/etc/rc.openvpn: OpenVPN: One or more OpenVPN tunnel endpoints may have changed its IP. Reloading endpoints that may use WAN_DHCP6.
Feb 12 20:31:57 OPNvirt opnsense: /usr/local/etc/rc.openvpn: OpenVPN: Resync server1 SpaceNet OpenVPN Server
```

My guess is that the restarts are too close to each other so  one get's started while the previous one still did not finish. And maybe the third one created the PID but then failed to start ...

**Comment by marjohn56** [comment]
#3229 possible fix ...  needs testing

**Comment by fichtner** [comment]
@BPtLNfxZWo @Space2Man can you please try c217bee ?

    # opnsense-patch c217bee

**Comment by fichtner** [comment]
Some patches from pfSense:

https://github.com/pfsense/pfsense/commit/8845e13
https://github.com/pfsense/pfsense/commit/ce98375

**Comment by BPtLNfxZWo** [comment]
Seems to work right now. 
GUI information correct, error messages gone, tunnel could be established
Thanks

**Comment by Space2Man** [comment]
Seems to work fine! After reboot service is listed as started!

**Comment by marjohn56** [comment]
Nice one Franco, I did do a quick scan bit missed it. Well done.

**Comment by fichtner** [comment]
Ok, I cannot guarantee that this will not fail later anyway. Patches posted earlier suggest that OpenVPN may be issuing its PID files too late, but we'll see when we get there.

Thanks everyone! ❤️

**Comment by BPtLNfxZWo** [comment]
Hi,
checked today, same issue again :-(

**Comment by marjohn56** [comment]
@Space2Man & @BPtLNfxZWo 

Looking at some debug info that I have been sent. Are you using a Dyn DNS service and are you using dhcp6 for WAN?

**Comment by BPtLNfxZWo** [comment]
Hi, in my case I'm using Dynamic DNS but no dhcpv6 on WAN.

**Comment by nivek1612** [comment]
@marjohn56 I'll be in France from Friday my set up there is Dynamic DNS and dhcp6c on WAN. Want me to test anything ?

**Comment by marjohn56** [comment]
Yes, everything!Sent from my Samsung Galaxy smartphone.<div>
</div><div>
</div><div>-------- Original message --------</div><div>From: Kev Willers <notifications@github.com> </div><div>Date: 20/02/2019  13:34  (GMT+00:00) </div><div>To: opnsense/core <core@noreply.github.com> </div><div>Cc: Martin Wasley <martin@queens-park.com>, Mention <mention@noreply.github.com> </div><div>Subject: Re: [opnsense/core] OpenVPN process status says not running when it is. (#3223) </div><div>
</div>@marjohn56 I'll be in France from Friday my set up there is Dynamic DNS and dhcp6c on WAN. Want me to test anything ?

**Comment by marjohn56** [comment]
> Hi, in my case I'm using Dynamic DNS but no dhcpv6 on WAN.

so no ipv6?

**Comment by BPtLNfxZWo** [comment]
Hi, I'm have ipv6 activated, but only for (internal) testing purposes and not on the WAN Interface.

**Comment by marjohn56** [comment]
Thanks, that rules something out, so no ipv6. Can you disable your Dyn DNS update and see if that has any effect on your openvpn issue. I'm trying to rule things out, I have no issues at all and I'm trying to replicate it.

**Comment by kkohio** [comment]
I just wanted to add that I am on a fresh setup on 19.1 LibreSSL side with very little configured. No IPv6 or DynDNS... just two WANs with failover, ipv4 DHCP on the LAN. I single port forwarding rule and pretty much everything else is pristine.

**Comment by Benqer0** [comment]
I have the same problem:


Mar 7 21:14:33 | openvpn[34459]: Exiting due to fatal error
-- | --
Mar 7 21:14:33 | openvpn[34459]: Cannot open TUN/TAP dev /dev/tun1: Device busy (errno=16)
Mar 7 21:14:33 | openvpn[34459]: TUN/TAP device ovpns1 exists previously, keep at program end
Mar 7 21:14:33 | openvpn[34459]: NOTE: the current --script-security setting may allow this configuration to call user-defined scripts
Mar 7 21:14:33 | openvpn[34138]: library versions: OpenSSL 1.0.2q 20 Nov 2018, LZO 2.10
Mar 7 21:14:33 | openvpn[34138]: OpenVPN 2.4.7 amd64-portbld-freebsd11.2 [SSL (OpenSSL)] [LZO] [LZ4] [MH/RECVDA] [AEAD] built on Feb 27 2019

OPNsense 19.1.2-amd64
FreeBSD 11.2-RELEASE-p9-HBSD
OpenSSL 1.0.2q 20 Nov 2018

**Comment by JasMan78** [comment]
Same issue here. It suddendly appeared after I've changed all my LAN subnet addresses by replacing them in the config backup file and restore it back.

OPNsense 19.1.4-amd64
Intel(R) Celeron(R) CPU J1900 @ 1.99GHz (4 cores)

**Comment by Space2Man** [comment]
FYI: I am on 19.1.4 and issue is back ... yes, I am using IPv6 on WAN but no DynDNS.

**Comment by Space2Man** [comment]
Funny, right after posting this 19.1.5 was released ... I just updated and OpenVPN is in status green again ... I will monitor.

Thanks!

**Comment by fichtner** [comment]
I'll be back eventually. This is a race condition inside OpenVPN and its PID file creation in which OpenVPN backgrounds itself before the PID is ready. Someone needs to write safeguard code here in order to fix this. I don't think OpenVPN will....

**Comment by Space2Man** [comment]
Hi, can confirm ... after "saving" WAN interface to trigger reload issue is back ... really seems to be some race condition. Nevertheless thanks for all the support ... I mean, it's working ... so it's only a display issue.

**Comment by fichtner** [comment]
I'll try to look at it for 19.7. At the moment, however, priorities lie elsewhere.

**Comment by Seethaar** [comment]
I tested this on a VM. When I built the new OPNsense firewall on version 19.7 (not the latest but just 19.7), there were no issues. But I took a VMware snapshot and spun a new firewall up from that template, which carried the CA, self-signed cert, VPN server and OTP server and I observed that the firewall is exhibiting the same behaviour mentioned on this issue.

1) Dashboard says OPENVPN not working.
2) Server error log : openvpn[66505]: Exiting due to fatal error
3) VPN server 'Connection Status' indeterminable.
4) Clients can connect successfully.

Thought worth sharing.
